File: app/windows/HyperWhisper/Services/LicenseUsageTracker.cs
// LICENSE USAGE TRACKER
// Tracks daily transcription time and model downloads for trial users.
//
// TODO-verify (Windows/CI): Rust shared-core swap. UNVERIFIED / compile-only.
//
// Wave 3 swap: all usage state + limit enforcement now lives in the `hw-license`
// Rust core, persisted via RustCoreKeyValueStore under com.hyperwhisper.usage.*.
// This class is a thin shim that keeps the public surface (singleton, properties,
// UsageChanged event, Record*/Can*/GetRemaining* methods, static TrialDaily*/
// TrialModel* display fields) and delegates to HyperwhisperCoreMethods.
//
// now-INJECTION: PLAIN UTC. Native Windows reset daily usage on the UTC calendar
// day (CheckDailyReset used DateTime.UtcNow.Date), which matches the core's
// `now/86400` UTC-day bucket exactly — so no local offset is applied (unlike
// macOS). The day-boundary reset is now read-time in the core (no midnight timer).
//
// TRIAL LIMITS: defaults 300s release / 1800s debug, 3 model downloads,
// overlaid with a fresh remote override (24h TTL) via RustLicenseCore.EffectiveLimits.

using System;
using HyperWhisper.Models;
using uniffi.hyperwhisper_core;

namespace HyperWhisper.Services;

/// <summary>
/// Tracks usage limits for trial users (delegated to the Rust core).
/// Licensed users have unlimited access.
/// </summary>
public sealed class LicenseUsageTracker
{
    // =========================================================================
    // CONSTANTS / DISPLAY LIMITS
    // =========================================================================

    // Display-only mirrors of the active limits (read directly by Settings UI).
    // Initialized from the core's build-flavor defaults; updated by
    // UpdateTrialLimits when a remote override is applied.

    public static int TrialDailyLimitSeconds { get; private set; } =
        (int)HyperwhisperCoreMethods.LicenseLimitsDefaults(RustLicenseCore.DebugBuild).dailySeconds;

    public static int TrialModelLimit { get; private set; } =
        (int)HyperwhisperCoreMethods.LicenseLimitsDefaults(RustLicenseCore.DebugBuild).modelDownloads;

    // =========================================================================
    // SINGLETON INSTANCE
    // =========================================================================

    private static LicenseUsageTracker? _instance;
    private static readonly object _lock = new();

    public static LicenseUsageTracker Instance
    {
        get
        {
            if (_instance == null)
            {
                lock (_lock)
                {
                    _instance ??= new LicenseUsageTracker();
                }
            }
            return _instance;
        }
    }

    // =========================================================================
    // STATE
    // =========================================================================

    private LicenseStatus _licenseStatus = LicenseStatus.Trial;

    /// <summary>Fired when usage statistics change.</summary>
    public event EventHandler? UsageChanged;

    private LicenseUsageTracker()
    {
        // Ensure the one-shot legacy migration has run before any usage read.
        _ = RustCoreKeyValueStore.Instance;
        var snap = Snapshot();
        LoggingService.Info(
            $"LicenseUsageTracker: Initialized (daily: {snap.dailySecondsUsed}s, models: {snap.modelsDownloaded})");
    }

    // =========================================================================
    // CORE HELPERS
    // =========================================================================

    private KeyValueStore Store => RustLicenseCore.Store;

    private UsageSnapshot Snapshot()
    {
        // TODO-verify (Windows/CI): Rust shared-core swap.
        return HyperwhisperCoreMethods.LicenseCheckLimits(
            Store, RustLicenseCore.ToCore(_licenseStatus), RustLicenseCore.EffectiveLimits(), RustLicenseCore.Now());
    }

    // =========================================================================
    // PUBLIC PROPERTIES
    // =========================================================================

    /// <summary>Daily transcription usage in seconds (resets at UTC day in core).</summary>
    public int DailyUsageSeconds => (int)Snapshot().dailySecondsUsed;

    /// <summary>Lifetime model-download count.</summary>
    public int ModelsDownloaded => (int)Snapshot().modelsDownloaded;

    /// <summary>Whether the daily limit has been reached (trial users only).</summary>
    public bool IsDailyLimitReached => Snapshot().dailyLimitReached;

    /// <summary>Whether the model download limit has been reached (trial users only).</summary>
    public bool IsModelLimitReached => Snapshot().modelLimitReached;

    // =========================================================================
    // LICENSE STATUS / LIMITS
    // =========================================================================

    /// <summary>
    /// Updates the display limits from a remote override and persists the override
    /// so the core applies it on every subsequent check.
    /// </summary>
    public void UpdateTrialLimits(int dailySeconds, int modelLimit)
    {
        TrialDailyLimitSeconds = dailySeconds;
        TrialModelLimit = modelLimit;

        // TODO-verify (Windows/CI): Rust shared-core swap.
        HyperwhisperCoreMethods.LicenseStoreRemoteOverride(
            Store,
            new TrialLimits(@dailySeconds: dailySeconds, @modelDownloads: modelLimit),
            RustLicenseCore.Now());

        LoggingService.Info($"LicenseUsageTracker: Trial limits updated (daily={dailySeconds}s, models={modelLimit})");
        NotifyUsageChanged();
    }

    /// <summary>Updates the license status (drives limit enforcement).</summary>
    public void UpdateLicenseStatus(LicenseStatus status)
    {
        var changed = _licenseStatus != status;
        _licenseStatus = status;

        if (changed)
        {
            LoggingService.Info($"LicenseUsageTracker: License status updated to {status}");
            NotifyUsageChanged();
        }
    }

    // =========================================================================
    // RECORDING LIMITS
    // =========================================================================

    /// <summary>Checks if user can start recording based on the daily limit.</summary>
    public bool CanStartRecording()
    {
        // TODO-verify (Windows/CI): Rust shared-core swap.
        return HyperwhisperCoreMethods.LicenseCanStartRecording(
            Store, RustLicenseCore.ToCore(_licenseStatus), RustLicenseCore.EffectiveLimits(), RustLicenseCore.Now());
    }

    /// <summary>Records transcription time and updates usage.</summary>
    public void RecordTranscriptionTime(int seconds)
    {
        // TODO-verify (Windows/CI): Rust shared-core swap.
        HyperwhisperCoreMethods.LicenseRecordUsage(Store, seconds, RustLicenseCore.Now());
        LoggingService.Info($"LicenseUsageTracker: Recorded {seconds}s transcription");
        NotifyUsageChanged();
    }

    /// <summary>Gets remaining daily transcription time in seconds (int.MaxValue if licensed).</summary>
    public int GetRemainingDailyTime()
    {
        var remaining = Snapshot().remainingDailySeconds;
        return remaining >= int.MaxValue ? int.MaxValue : (int)remaining;
    }

    /// <summary>Gets remaining daily time as a formatted string.</summary>
    public string GetRemainingDailyTimeFormatted()
    {
        var remaining = GetRemainingDailyTime();
        if (remaining == int.MaxValue)
            return "Unlimited";

        var minutes = remaining / 60;
        var seconds = remaining % 60;
        return minutes > 0 ? $"{minutes}m {seconds}s" : $"{seconds}s";
    }

    // =========================================================================
    // MODEL DOWNLOAD LIMITS
    // =========================================================================

    /// <summary>Checks if user can download another model.</summary>
    public bool CanDownloadModel()
    {
        // TODO-verify (Windows/CI): Rust shared-core swap.
        return HyperwhisperCoreMethods.LicenseCanDownloadModel(
            Store, RustLicenseCore.ToCore(_licenseStatus), RustLicenseCore.EffectiveLimits());
    }

    /// <summary>Increments the lifetime model download count.</summary>
    public void IncrementModelDownloadCount()
    {
        // TODO-verify (Windows/CI): Rust shared-core swap.
        HyperwhisperCoreMethods.LicenseRecordModelDownload(Store);
        LoggingService.Info($"LicenseUsageTracker: Model download recorded (total: {ModelsDownloaded})");
        NotifyUsageChanged();
    }

    /// <summary>Gets remaining model downloads (int.MaxValue if licensed).</summary>
    public int GetRemainingModelDownloads()
    {
        var remaining = Snapshot().remainingModelDownloads;
        return remaining >= int.MaxValue ? int.MaxValue : (int)remaining;
    }

    // =========================================================================
    // PRIVATE
    // =========================================================================

    private void NotifyUsageChanged()
    {
        UsageChanged?.Invoke(this, EventArgs.Empty);
    }
}

File: app/macos/hyperwhisper/Managers/LicenseUsageTracker.swift
//
//  LicenseUsageTracker.swift
//  hyperwhisper
//
//  LICENSE USAGE TRACKING
//  Tracks daily transcription time and model downloads for trial users.
//
//  TRIAL LIMITS:
//  - Daily transcription: 300 seconds (5 minutes) in production, 1800 seconds (30 minutes) in dev
//  - Model downloads: 3 models maximum
//
//  LICENSED USERS:
//  - No limits enforced (Int.max for all checks)
//
//  STORAGE (M3-C):
//  - All usage/limit LOGIC now lives in the Rust shared core (`hw-license`).
//  - State is persisted in UserDefaults via a shared `RustLicenseStore`
//    (`com.hyperwhisper.usage.*`), seeded once from the legacy Core Data
//    `UsageTracking` entity (see RustLicenseStore.swift).
//  - The day-boundary reset is READ-TIME in the core (no midnight timer needed);
//    `now` is passed offset to the local day so the UTC bucket lands on local
//    midnight, preserving the old `isDateInToday` behavior.
//
//  INTEGRATION:
//  - Called by AudioRecordingManager after each transcription
//  - Called by WhisperModelManager after model downloads
//  - Queried by UI to show remaining time/downloads
//

import Foundation
import SwiftUI
import AppKit

/// Tracks usage limits for trial users.
///
/// This class is a thin macOS shim over the Rust license core: it owns the
/// `@Published` UI state and forwards every decision (record/check/can-start) to
/// the core over the shared `RustLicenseStore`.
///
/// Licensed users bypass all limits and get unlimited access.
@MainActor
class LicenseUsageTracker: ObservableObject {

    // MARK: - Published Properties (for UI binding)

    /// Daily transcription usage in seconds (driven by the core's UsageSnapshot).
    @Published var dailyUsageSeconds: Int = 0

    /// Number of models downloaded by the user. Lifetime count, never resets.
    @Published var modelsDownloaded: Int = 0

    /// Whether the daily limit has been reached. true = cannot start recordings.
    @Published var isDailyLimitReached: Bool = false

    /// Whether the model download limit has been reached. true = cannot download.
    @Published var isModelLimitReached: Bool = false

    // MARK: - Limits (effective trial limits in force)

    /// Effective daily trial limit (defaults overlaid with a fresh remote
    /// override). Exposed for UI display. Computed from the core, not hardcoded.
    private(set) var trialDailyLimitSeconds: Int

    /// Effective trial model-download limit. Computed from the core.
    private(set) var trialModelLimit: Int

    // MARK: - Properties

    /// Shared key-value store backing the Rust license core.
    private let store: RustLicenseStore

    /// Current license status (determines if limits apply).
    private var licenseStatus: LicenseStatus = .trial

    /// `true` for DEBUG builds — drives the core's `licenseLimitsDefaults`.
    /// Passed in (not read with `#if DEBUG` here) so the default limit values
    /// come from the core, never hardcoded.
    private let debugBuild: Bool

    /// Token for the system wake observer, removed on deinit.
    /// Kept ONLY as a UI refresh trigger (re-reads the core snapshot). The
    /// midnight reset is now read-time in the core, so no reset timer is needed.
    private var wakeObserver: NSObjectProtocol?

    // MARK: - Public API (Read-only limits)

    /// Exposes the trial transcription limit for UI display.
    var trialDailyTranscriptionLimit: Int { trialDailyLimitSeconds }

    /// Exposes the trial model download limit for UI display.
    var trialModelDownloadLimit: Int { trialModelLimit }

    // MARK: - Initialization

    init(store: RustLicenseStore) {
        self.store = store
        #if DEBUG
        self.debugBuild = true
        #else
        self.debugBuild = false
        #endif

        // Seed effective limits from the core's defaults for this build flavor.
        let defaults = licenseLimitsDefaults(debugBuild: debugBuild)
        self.trialDailyLimitSeconds = Int(defaults.dailySeconds)
        self.trialModelLimit = Int(defaults.modelDownloads)

        // Refresh UI state when the Mac wakes from sleep. The core resets the
        // daily bucket at read time, so this is purely a UI re-read — no reset
        // timer is scheduled anymore.
        setupWakeObserver()
    }

    deinit {
        if let wakeObserver {
            NSWorkspace.shared.notificationCenter.removeObserver(wakeObserver)
        }
    }

    // MARK: - Effective limits

    /// The effective `Limits` in force right now: build-flavor defaults overlaid
    /// with a fresh remote override (if present and within the 24h TTL). This is
    /// the single source of truth passed to every core usage call.
    private func effectiveLimits() -> Limits {
        let defaults = licenseLimitsDefaults(debugBuild: debugBuild)
        if let override = licenseRemoteOverrideIfFresh(
            store: store,
            nowUnixSecs: RustLicenseTime.nowUTC()
        ) {
            return Limits(
                dailySeconds: override.dailySeconds,
                modelDownloads: override.modelDownloads
            )
        }
        return defaults
    }

    // MARK: - Remote Config Updates

    /// Applies freshly-fetched remote trial limits: persists them to the core's
    /// store (so they survive restarts and feed `effectiveLimits()`), updates the
    /// displayed limits, and recomputes the published flags.
    /// Called by LicenseManager after fetching remote config.
    func updateTrialLimits(dailySeconds: Int, modelLimit: Int, maxAgeSecs: Int? = nil, isLiveFetch: Bool = false) {
        licenseStoreRemoteOverride(
            store: store,
            limits: TrialLimits(
                dailySeconds: Int64(dailySeconds),
                modelDownloads: Int64(modelLimit)
            ),
            nowUnixSecs: RustLicenseTime.nowUTC()
        )
        // B4/E5: persist the server Cache-Control max-age so the core's freshness
        // check (`licenseRemoteOverrideIfFresh`) honors it instead of the fixed
        // default. Key must match hw_license::cache's K_OVERRIDE_MAX_AGE.
        //
        // Only a LIVE fetch is authoritative for the TTL. On a live fetch with a
        // positive max-age we store it; with no/≤0 max-age we CLEAR the key (the
        // core treats empty as "use the 6h default") so a previously-stored
        // max-age can't persist after the server stops sending one. The
        // cached-replay path (isLiveFetch == false) leaves the stored value
        // untouched — it isn't a fresh signal from the server.
        if isLiveFetch {
            if let maxAgeSecs, maxAgeSecs > 0 {
                store.set(key: "com.hyperwhisper.config.maxAgeSecs", value: String(maxAgeSecs))
            } else {
                store.set(key: "com.hyperwhisper.config.maxAgeSecs", value: "")
            }
        }
        trialDailyLimitSeconds = dailySeconds
        trialModelLimit = modelLimit
        refreshFromSnapshot()
    }

    // MARK: - License Status Updates

    /// Updates the license status and recomputes the published flags.
    func updateLicenseStatus(_ status: LicenseStatus) {
        licenseStatus = status
        refreshFromSnapshot()
    }

    // MARK: - Recording Limits

    /// Checks if the user can start recording based on the daily limit.
    /// Delegates to the core (which day-resets at read time).
    func canStartRecording() -> Bool {
        let allowed = licenseCanStartRecording(
            store: store,
            status: LicenseNetworkService.toCore(licenseStatus),
            limits: effectiveLimits(),
            nowUnixSecs: RustLicenseTime.nowLocal()
        )
        // Keep the published flags consistent with what we just decided.
        refreshFromSnapshot()
        return allowed
    }

    /// Records transcription time against today's bucket, then refreshes the UI.
    func recordTranscriptionTime(_ seconds: Int) async {
        licenseRecordUsage(
            store: store,
            seconds: Int64(seconds),
            nowUnixSecs: RustLicenseTime.nowLocal()
        )
        refreshFromSnapshot()
    }

    /// Remaining daily transcription time in seconds (Int.max for licensed users).
    func getRemainingDailyTime() -> Int {
        let snap = currentSnapshot()
        return clampRemaining(snap.remainingDailySeconds)
    }

    // MARK: - Model Download Limits

    /// Checks if the user can download another model. Delegates to the core.
    func canDownloadModel() -> Bool {
        licenseCanDownloadModel(
            store: store,
            status: LicenseNetworkService.toCore(licenseStatus),
            limits: effectiveLimits()
        )
    }

    /// Increments the lifetime model download count, then refreshes the UI.
    func incrementModelDownloadCount() async {
        licenseRecordModelDownload(store: store)
        refreshFromSnapshot()
    }

    /// Remaining model downloads (Int.max for licensed users).
    func getRemainingModelDownloads() -> Int {
        let snap = currentSnapshot()
        return clampRemaining(snap.remainingModelDownloads)
    }

    // MARK: - Snapshot refresh

    /// Reads the authoritative usage snapshot from the core. The core performs
    /// the read-time day-boundary reset, so this always reflects today.
    private func currentSnapshot() -> UsageSnapshot {
        licenseCheckLimits(
            store: store,
            status: LicenseNetworkService.toCore(licenseStatus),
            limits: effectiveLimits(),
            nowUnixSecs: RustLicenseTime.nowLocal()
        )
    }

    /// Drives every `@Published` property from a fresh core snapshot.
    private func refreshFromSnapshot() {
        let snap = currentSnapshot()
        dailyUsageSeconds = Int(snap.dailySecondsUsed)
        modelsDownloaded = Int(snap.modelsDownloaded)
        isDailyLimitReached = snap.dailyLimitReached
        isModelLimitReached = snap.modelLimitReached
    }

    /// Refreshes usage statistics from the core (called on app launch / wake).
    /// Replaces the old Core Data reload; the core is now the source of truth.
    func refreshUsageStats() async {
        refreshFromSnapshot()
    }

    /// Maps the core's `i64::MAX` "unlimited" sentinel to Swift `Int.max`.
    private func clampRemaining(_ value: Int64) -> Int {
        value == Int64.max ? Int.max : Int(value)
    }

    // MARK: - Wake observer (UI refresh only)

    /// Observes system wake to refresh the UI snapshot after sleep.
    ///
    /// The core resets the daily bucket at read time, so a Mac that sleeps across
    /// midnight surfaces a fresh counter on the next read automatically. This
    /// observer just nudges the published state so the UI updates without a user
    /// action. No reset timer is scheduled.
    private func setupWakeObserver() {
        wakeObserver = NSWorkspace.shared.notificationCenter.addObserver(
            forName: NSWorkspace.didWakeNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            Task { @MainActor in
                self?.refreshFromSnapshot()
            }
        }
    }
}
